File: solutions/javascript/armstrong-numbers/3/armstrong-numbers.js

export function isArmstrongNumber(number) {

  const digits = number.toString().split('').map(d => Number.parseInt(d));
  let powerSum;
  if (digits.length < 16) {
    const exp = digits.length;
    powerSum = digits.reduce((acc,val) => acc+val**exp, 0);
  } else {
    const exp = BigInt(digits.length);
    powerSum = digits.reduce((acc,val) => acc+BigInt(val)**exp, 0n);
  }
  return number == powerSum;
};

File: solutions/swift/armstrong-numbers/1/Sources/ArmstrongNumbers/ArmstrongNumbers.swift
func isArmstrongNumber(_ number: Int) -> Bool {
    let digits = toReversedDigits(number)
    let sum = digits.reduce(0) { (acc, digit) in acc + pow(digit, to: digits.count) }
    return number == sum
}

// Note: this returns the digits in reverse order but it doesn't matter
// when checking if a number is an armstorng number
func toReversedDigits(_ number: Int) -> [Int] {
    var digits = [Int]()
    var remaining = number
    while remaining > 0 {
        digits.append(remaining % 10)
        remaining /= 10
    }
    return digits
}

func pow(_ n: Int, to: Int) -> Int {
    return (1...to).reduce(1) { acc, _ in acc * n }
}
